### PR TITLE
Refactor TravisCI and docker integration

### LINKED ISSUE #1681: Optimize the CI and docker settings
Few ideas to improve the CI pipeline:
 * Looks like our CI is building docker images and not reusing any already available image.
 * we compile the PHP x-debug extension, while it is not used during the tests
 * running the php-cs-fixer could be done in travis-ci, no docker images are needed for that

### PR DESCRIPTION
**Pull request by thePanz**
Closes #1681 

Note: The CI running time went from ~19minutes to ~4minutes

Todo:
- [x] Enable Travis for all PHP and ES versions
- [x] Ensure local tests are green
- [x] Use NGINX for proxy tests
- [x] Add changelog (plus note about updated libraries)

### CODE REVIEW

**Comment by thePanz** [comment]
@ruflin this is a first try to use the default ES docker images, and rely on travis' PHP.
I am using phive to pull in the "dev" tools, such as phpunit and php-cs-fixer

**Comment by p365labs** [comment]
@thePanz & @ruflin My only question about this implementation is how to execute locally tests, without Docker, we cannot guarantee to execute tests with the same configuration and with the same condition of CI.

**Comment by thePanz** [comment]
@p365labs docker is still available, I have mainly working on simplifying the ES containers, by using the existing images, instead of building them every time on Travis.

For locally running the tests, I'd like to propose the following options
1.  require the developer to have (at least) PHP locally available, then running the tests locally: spinning up ES in Docker  by using the docker-compose files available in the repo (and used by the CI, too)
2. provide both ES and PHP containers, and run the tests from the PHP container (this would help the dev, in case the local PHP version is not compatible with the min-requirements of the library... but I guess this is quite rare?)

I like the first option more: Elastica is a library and contributors might use any (supported) PHP version to work on it. wdyt?

**Comment by thePanz** [comment]
Question for @ruflin : do we really need the "proxy" tests? looks like we're testing Guzzle, instead of the Elastica library... any inputs?

**Comment by p365labs** [comment]
@thePanz got your point, but in general I prefer not to have or to use PHP locally. Docker executed locally assures you to have exactly the PHP version you need, the specified ES version you need.  Also I'm working with many different libraries and sometomes they need to have specific php extensions.
It assures idempotency of your test environments.

I'm using a MAC and I don't really know which version to use, in general the approach of testing with Docker also locally assures me to have the same environment ready for tests with all thhat we need.

**Comment by thePanz** [comment]
@p365labs I see. I'll add a PHP container, able to run tests in it then (option 2 from my previous message). I'll use the "lowest" php version supported by the Elastica library.

I'll add more "make" commands to run those tests on the php+es containers, too


wdyt?

**Comment by ruflin** [comment]
Thanks @thePanz for tackling this. As you know I'm torn around this change. The history of all this setup is, that I wanted everyone running EXACTLY the same tests, independent of the user, platform CI it is running on. I've spent a lot of time in the past figuring out why some tests fail locally for someone but don't fail on my end, or they fail on CI and not my system. Especially if things fail on CI and not locally, things get hairy to debug.

I'm ok with most changes in this PR and +1 on being still able to run it locally in a Docker container. The question that stays open for me is: Should Travis run on Docker or not for all the above reasons.

@thePanz To keep things moving: In the second part of your PR, you sneaked in lots of code changes / cleanups. Any chance you could open a separate PR for these to get them in?

**Comment by thePanz** [comment]
@ruflin I extracted the code-changes into https://github.com/ruflin/Elastica/pull/1742

And I agree about your concerns on different PHP versions. I'll re-add the possibility to run the tests on a dedicated PHP container. The different PHP versions issues you talked about: do you remember those ones? If those were related to missing dependencies, we'd better check our composer.json to require them.

And yes, I am of the opinion that Travis should use Docker, at least for the ES instances. :)

**Comment by thePanz** [comment]
@ruflin @p365labs I added a PHP container, which can be used to run `phpunit` and `php-cs-fixer`.
The details are in the `contributing.md` file.

The local docker uses `php:7.2` so that we're on the lowest supported version of PHP. Tested locally and it worked fine.

The default ES version that is created when running `make docker-up` is 7.2, but can be dynamically changed too.

**Comment by p365labs** [comment]
Tests are failing, locally, only 14 and are about this error
```Connection to localhost:12345 failed```
which is related to PROXY_HOST env var and nginx used as a proxy

**Comment by p365labs** [comment]
I've noticed that you amend a lot of tests, as this PR is about CI+Docker can you split up and open another PR with this files?

Screenshot 2020-01-16 at 06 46 58

**Comment by thePanz** [comment]
> @p365labs : I've noticed that you amend a lot of tests, as this PR is about CI+Docker can you split up and open another PR with this files?

The PR is already available here #1742. I will rebase this PR on master as soon as it is merged

File: test/Elastica/SnapshotTest.php [hunk 1]
```diff
@@ -12,7 +12,7 @@
  */
 class SnapshotTest extends Base
 {
-    private const SNAPSHOT_PATH = '/usr/share/elasticsearch/data/';
+    private const SNAPSHOT_PATH = '/usr/share/elasticsearch/repository/';
```
Comment (thePanz) [inline test/Elastica/SnapshotTest.php]: This is required as the storage location of the repository changed: we use 2 different volumes for the ES data and the repository (the latter is shared among all nodes)

**Comment by thePanz** [comment]
> good, I did not noticed the #1742, so you can remove that amendings from this PR, right?

@p365labs done, the changes in the tests are now related only to the changes introduced by the Docker containers.

File: Makefile [hunk 2]
```diff
@@ -1,184 +1,103 @@
...
+run-phpdoc: tools/phpdocumentor.phar
+	tools/phpdocumentor.phar --directory=lib --target=build/docs --template=clean
+
+##
+## Docker commands
+##
+
+.PHONY: docker-start
+docker-start:
+	sudo sysctl -w vm.max_map_count=262144; \
```
[the remark this answers is not in the document]
Comment (thePanz) [reply Makefile]: I removed the setting from the Makefile, and added a note in the CONTRIBUTING file

File: .gitignore [hunk 3]
```diff
@@ -1,17 +1,9 @@
 .DS_Store
 .idea
-*.iml
-.vagrant
 .php_cs
 .php_cs.cache
 
-cache.properties
-/nbproject
-build
```
[the remark this answers is not in the document]
Comment (thePanz) [reply .gitignore]: Added /build to the ignored files

File: docker/docker-compose.es71.yml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+version: '3.4'
```
[the remark this answers is not in the document]
Comment (thePanz) [reply docker/docker-compose.es71.yml]: I sometimes found that in minor ES versions some APIs changed (or some bugs were fixed).
I would like the Elastica library not t rely on edge-cases of specific minor-versions.

If you believe that we should only test the library on one ES version let me know

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,96 +1,95 @@
...
 
+This repository comes with a set of docker-compose.yml templates used for the CI and the local development.
 
-Setup
------
-Elastica uses docker for it's development environment. To use the docker environment, check out the Makefile for the necessary commands.
+To start the local docker instance run `make docker-start`: the command will pull the required containers for PHP and ES.
+For ES to properly run, the `vm.max_map_count=262144` System configuration is needed by ES to properly spin up the nodes.
+
+Check out the Makefile for other commands that can be used to run tests and other operations:
```
[the remark this answers is not in the document]
Comment (thePanz) [reply CONTRIBUTING.md]: Added in the CONTRIBUTING file

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (p365labs) [reply Makefile]: yep seems legit, on OSX it's a bit tricky to set it up. I've not yet found a good way to set it up. the only way I know  right now is this

```On Latest Docker For Mac (Version 18.06.0-ce-mac70):
$ screen ~/Library/Containers/com.docker.docker/Data/vms/0/tty
On the blank screen, press RETURN.

Then configure the sysctl setting as you would for Linux:

sysctl -w vm.max_map_count=262144
Exit by Control-A Control-\
```
some more infos [HERE](https://stackoverflow.com/questions/41192680/update-max-map-count-for-elasticsearch-docker-container-mac-host)

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,96 +1,95 @@
...
-```
-will do a check executing php-cs-fixer in ```dry-run``` mode. It is usually executed during Travis build.
+### Tests
+Run the command `make docker-run-phpunit` to run the PHP tests on the code inside the docker container.
 
-  Exit codes
-  ----------
+If you want to run just a specific test or a one specific file, you should run the `tools/phpunit.phar` from inside
+the docker container:
+    1. enter the docker container `docker exec -ti 'elastica_php' sh`
```
[the remark this answers is not in the document]
Comment (thePanz) [reply CONTRIBUTING.md]: Added, and documented in CONTRIBUTING file.
Example: `make docker-run-phpunit PHPUNIT_OPTIONS="test/Elastica/ClientTest.php"`

File: Makefile [hunk 7]
```diff
@@ -1,184 +1,103 @@
...
+	docker-compose --file=docker/docker-compose.yml --file=docker/docker-compose.es.yml --file=docker/docker-compose.es${START_ES_VERSION}.yml up --build
+
+.PHONY: docker-stop
+docker-stop:
+	docker-compose --file=docker/docker-compose.yml --file=docker/docker-compose.es.yml --file=docker/docker-compose.es${START_ES_VERSION}.yml down
+
+.PHONY: docker-run-phpunit
+docker-run-phpunit:
```
[the remark this answers is not in the document]
Comment (thePanz) [reply Makefile]: I documented it in the CONTRIBUTING: as a developer you should keep an eye to the ES logs, right? :)
My solution: run `make docker-start` in one terminal, and run `make docker-run-phpunit` in another. The ES logs will be visible in the first terminal, including errors and logs in case one of the containers fail to start.

In this was the following error could be found and fixed in any local environment ;-)
```
es02_1  | ERROR: [1] bootstrap checks failed
es02_1  | [1]: max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]
```

**Comment by thePanz** [comment]
@ruflin @p365labs I was able to get all "proxy" tests working with the new Docker setup, build is green on both local and Travis.

Would you like to have another review?

**Comment by thePanz** [comment]
I had to update few more things to have it working with the new tests (namely: PHP 7.4 and the `--prefer-lower` composer flag):

    - Update elastica-php library, 7.1.1 fixes an issue on Ingestion/Put() typehinting
    - Update guzzle to 6.3.0, fixes issue with PHP >= 7.2
    - Update phpunit to 7.5.x, compatible with PHP 7.4

File: test/bootstrap.php [hunk 8]
```diff
@@ -1,5 +0,0 @@
-<?php
-
-\defined('BASE_PATH') || \define('BASE_PATH', \realpath(__DIR__));
-
-require \dirname(__DIR__).'/vendor/autoload.php';
```
Comment (ruflin) [inline test/bootstrap.php]: Can't comment on the rename of the file below, but I wonder why it still works?

File: .gitignore [hunk 9]
```diff
@@ -1,17 +1,6 @@
-.DS_Store
```
[the remark this answers is not in the document]
Comment (thePanz) [reply .gitignore]: see: https://twitter.com/Ocramius/status/1213219735271362561 :)
> A .gitignore in a project should only cover artifacts caused by the contained source code, not those caused by the person

File: test/bootstrap.php (hunk 8 above)
Comment (thePanz) [reply test/bootstrap.php]: phpunit can load the `vendor/bootstrap.php` file automatically ;)

**Comment by thePanz** [comment]
@ruflin yes, chatting it live would be more efficient :)
Should we organize an PHP-Elastica hack-day in ZH to work on the top-priority PR/issues?

**Comment by thePanz** [comment]
@ruflin any updates on this PR? :)

File: .gitignore (hunk 9 above)
[the remark this answers is not in the document]
Comment (thePanz) [reply .gitignore]: I created: https://github.com/ruflin/Elastica/pull/1753 to handle that :)

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,96 +1,121 @@
...
-are not respected **build will fail**. 
-No local setup of the tools is needed as everything will happen directly in the container.
+## Commands
+The advantage in using the commands below is that no local tools and libraries have to be installed and it is guaranteed
+that everyone is using the same tools.
 
-```
-make lint
-```
+The tools required for each command will be installed only when needed byt the command itself.
```
[the remark this answers is not in the document]
Comment (thePanz) [reply CONTRIBUTING.md]: Fixed

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,96 +1,121 @@
...
-  * 64 Exception raised within the application.
+Run the command `make docker-run-phpunit` to run the PHP tests on the code inside the docker container.
 
-Coding
-------
+Options can be passed to `phpunit` when running in the docker container: use the `PHPUNIT_OPTIONS` to pass additional
+arguments to the invocation of the tool.
+
+Examples:
+ - run a specific group of tests: `make docker-run-phpunit PHPUNIT_OPTIONS="--group=unit"`
```
[the remark this answers is not in the document]
Comment (thePanz) [reply CONTRIBUTING.md]: Can you first run `make clean` to ensure that all libraries and cached files are removed?

File: .gitignore [hunk 12]
```diff
@@ -6,11 +6,14 @@
...
 
 cache.properties
-/nbproject
-build
 atlassian-ide-plugin.xml
 .settings
+
+/build
+/nbproject
+/tools
```
Comment (ruflin) [inline .gitignore]: No need to change but I wonder if /tools should go under /build?

File: CONTRIBUTING.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply CONTRIBUTING.md]: That worked!
